File: mobile/src/main/java/com/example/steven/stlbusarrivals/XmlParser.java
package com.example.steven.stlbusarrivals;

import android.os.AsyncTask;

import com.example.steven.stlbusarrivals.model.Path;
import com.example.steven.stlbusarrivals.model.PathBounds;
import com.example.steven.stlbusarrivals.model.Point;
import com.example.steven.stlbusarrivals.model.PathList;
import com.example.steven.stlbusarrivals.model.Route;
import com.example.steven.stlbusarrivals.model.RouteList;
import com.example.steven.stlbusarrivals.model.Stop;
import com.example.steven.stlbusarrivals.model.StopList;
import com.example.steven.stlbusarrivals.model.TimeList;
import com.example.steven.stlbusarrivals.model.TimePrediction;
import com.example.steven.stlbusarrivals.model.Vehicule;
import com.example.steven.stlbusarrivals.model.VehiculeList;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserFactory;

import java.io.StringReader;
import java.util.Observable;

/**
 * Created by Steven on 2016-01-25.
 */
public class XmlParser extends Observable{

    public void readXml(final int xmlType, final String xml){
        new AsyncTask<Void, Void, Void>(){
            RouteList routeList;
            StopList stopList;
            PathList pathList;
            PathBounds pathBounds;
            TimeList timeList;
            VehiculeList vehiculeList;
            @Override
            protected Void doInBackground(Void... params) {
                try {
                    XmlPullParserFactory factory = XmlPullParserFactory.newInstance();
                    factory.setNamespaceAware(true);
                    XmlPullParser xpp = factory.newPullParser();

                    xpp.setInput(new StringReader(xml));
                    int eventType = xpp.getEventType();
                    switch(xmlType){
                        case Constants.ROUTE_XML:
                            routeList = new RouteList();
                            while (eventType != XmlPullParser.END_DOCUMENT) {
                                if (eventType == XmlPullParser.START_TAG) {
                                    if (xpp.getName().equals("route")) {
                                        Route route = new Route();
                                        route.setTag(xpp.getAttributeValue(0));
                                        route.setTitle(xpp.getAttributeValue(1));
                                        routeList.add(route);
                                    }
                                }
                                eventType = xpp.next();
                            }
                            return null;
                        case Constants.STOP_XML:
                            stopList = new StopList();
                            pathList = new PathList();
                            pathBounds = new PathBounds();
                            while (eventType != XmlPullParser.END_DOCUMENT) {
                                if (eventType == XmlPullParser.START_TAG) {
                                    if(xpp.getName().equals("route")){
                                        pathBounds.setLatMin(Double.valueOf(xpp.getAttributeValue(4)));
                                        pathBounds.setLatMax(Double.valueOf(xpp.getAttributeValue(5)));
                                        pathBounds.setLongMin(Double.valueOf(xpp.getAttributeValue(6)));
                                        pathBounds.setLongMax(Double.valueOf(xpp.getAttributeValue(7)));
                                    }
                                    if (xpp.getName().equals("stop") && xpp.getAttributeCount() > 1) {
                                        Stop stop = new Stop();
                                        stop.setTag(xpp.getAttributeValue(0));
                                        stop.setTitle(xpp.getAttributeValue(1));
                                        stop.setId(xpp.getAttributeValue(4));
                                        stopList.add(stop);
                                    }
                                    if(xpp.getName().equals("path")){
                                        Path path = new Path();
                                        while(eventType != XmlPullParser.END_TAG || (xpp.getName() !=null && xpp.getName().equals("point"))) {
                                            if(xpp.getName() != null && xpp.getName().equals("point")) {
                                                Point point = new Point();
                                                point.setLatitude(Double.valueOf(xpp.getAttributeValue(0)));
                                                point.setLongitude(Double.valueOf(xpp.getAttributeValue(1)));
                                                path.add(point);
                                            }
                                            eventType = xpp.next();
                                        }
                                        pathList.add(path);
                                    }
                                }
                                eventType = xpp.next();
                            }
                            return null;
                        case Constants.PREDICTION_XML:
                            timeList = new TimeList();
                            while (eventType != XmlPullParser.END_DOCUMENT) {
                                if (eventType == XmlPullParser.START_TAG) {
                                    if (xpp.getName().equals("prediction")) {
                                        TimePrediction timePrediction = new TimePrediction();
                                        timePrediction.setTime(xpp.getAttributeValue(2));
                                        timeList.add(timePrediction);
                                    }
                                }
                                eventType = xpp.next();
                            }
                            return null;
                        case Constants.LOCATION_XML:
                            vehiculeList = new VehiculeList();
                            while (eventType != XmlPullParser.END_DOCUMENT) {
                                if (eventType == XmlPullParser.START_TAG) {
                                    if (xpp.getName().equals("vehicle")) {
                                        Vehicule vehicule = new Vehicule();
                                        vehicule.setLongitude(xpp.getAttributeValue(4));
                                        vehicule.setLatitude(xpp.getAttributeValue(3));
                                        vehiculeList.add(vehicule);
                                    }
                                }
                                eventType = xpp.next();
                            }
                            return null;
                        default:
                            return null;
                    }
                }catch(Exception e){
                    e.printStackTrace();
                }
                return null;
            }

            @Override
            protected void onPostExecute(Void aVoid) {
                super.onPostExecute(aVoid);
                switch(xmlType){
                    case Constants.ROUTE_XML:
                        notifyObs(routeList);
                        break;
                    case Constants.STOP_XML:
                        notifyObs(stopList);
                        notifyObs(pathList);
                        notifyObs(pathBounds);
                        break;
                    case Constants.PREDICTION_XML:
                        notifyObs(timeList);
                        break;
                    case Constants.LOCATION_XML:
                        notifyObs(vehiculeList);
                        break;
                    default:
                        break;
                }
            }
        }.execute();
    }

    public void notifyObs(Object o){
        setChanged();
        notifyObservers(o);
    }
}

File: mobile/src/main/java/com/tilly/steven/stlbusarrivals/XmlParser.kt
package com.tilly.steven.stlbusarrivals

import android.util.Log
import androidx.core.text.HtmlCompat
import androidx.core.text.HtmlCompat.FROM_HTML_MODE_LEGACY
import com.tilly.steven.stlbusarrivals.model.*
import com.tilly.steven.stlbusarrivals.network.NetworkCallback
import kotlinx.coroutines.async
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import org.xmlpull.v1.XmlPullParserFactory
import java.io.IOException
import java.io.StringReader
import java.util.*

/**
 * Created by Steven on 2016-01-25.
 */
class XmlParser : Observable() {

    @Throws(XmlPullParserException::class, IOException::class)
    fun readRouteXml(xml: String, callback: NetworkCallback){
        launchUI {
            val list = async {
                val routeList = RouteList()
                val factory = XmlPullParserFactory.newInstance()
                factory.isNamespaceAware = true
                val xpp = factory.newPullParser()

                xpp.setInput(StringReader(xml))
                var eventType = xpp.eventType
                while (eventType != XmlPullParser.END_DOCUMENT) {
                    if (eventType == XmlPullParser.START_DOCUMENT) {
                    } else if (eventType == XmlPullParser.START_TAG) {
                        if (xpp.name == "route") {
                            val route = Route()
                            route.tag = xpp.getAttributeValue(0)
                            route.title = xpp.getAttributeValue(1)
                            routeList.add(route)
                        }
                    }
                    eventType = xpp.next()
                }
                routeList
            }.await()
            callback.onRouteListLoaded(list)
        }
    }

    @Throws(XmlPullParserException::class, IOException::class)
    fun readStopXml(xml: String, callback: NetworkCallback) {
        launchUI {
            val stopList = StopList()
            val pathList = PathList()
            val pathBounds = PathBounds()

            val triple = async {
                val factory = XmlPullParserFactory.newInstance()
                factory.isNamespaceAware = true
                val xpp = factory.newPullParser()

                xpp.setInput(StringReader(xml))
                var eventType = xpp.eventType
                while (eventType != XmlPullParser.END_DOCUMENT) {
                    if (eventType == XmlPullParser.START_DOCUMENT) {
                    } else if (eventType == XmlPullParser.START_TAG) {
                        if (xpp.name == "route") {
                            pathBounds.latMin = java.lang.Double.valueOf(xpp.getAttributeValue(4))
                            pathBounds.latMax = java.lang.Double.valueOf(xpp.getAttributeValue(5))
                            pathBounds.longMin = java.lang.Double.valueOf(xpp.getAttributeValue(6))
                            pathBounds.longMax = java.lang.Double.valueOf(xpp.getAttributeValue(7))
                        }
                        if (xpp.name == "stop" && xpp.attributeCount > 1) {
                            val stop = Stop()
                            stop.tag = xpp.getAttributeValue(0)
                            stop.title = xpp.getAttributeValue(1)
                            stop.id = xpp.getAttributeValue(4)
                            stopList.add(stop)
                        }
                        if (xpp.name == "path") {
                            val path = Path()
                            while (eventType != XmlPullParser.END_TAG || xpp.name != null && xpp.name == "point") {
                                if (xpp.name != null && xpp.name == "point") {
                                    val point = Point()
                                    point.latitude = java.lang.Double.valueOf(xpp.getAttributeValue(0))
                                    point.longitude = java.lang.Double.valueOf(xpp.getAttributeValue(1))
                                    path.add(point)
                                }
                                eventType = xpp.next()
                            }
                            pathList.add(path)
                        }
                    }
                    eventType = xpp.next()
                }
                Log.d("doInBackground", "finished")
                Triple(stopList, pathList, pathBounds)
            }.await()
            callback.onStopListLoaded(triple.first)
            callback.onPathListLoaded(triple.second)
            callback.onPathBoundsLoaded(triple.third)
        }
    }

    @Throws(XmlPullParserException::class, IOException::class)
    fun readPrediction(xml: String) {
        launchUI{
            val timeList = TimeList()

            async {
                val factory = XmlPullParserFactory.newInstance()
                factory.isNamespaceAware = true
                val xpp = factory.newPullParser()

                xpp.setInput(StringReader(xml))
                var eventType = xpp.eventType
                while (eventType != XmlPullParser.END_DOCUMENT) {
                    if (eventType == XmlPullParser.START_DOCUMENT) {
                    } else if (eventType == XmlPullParser.START_TAG) {
                        if (xpp.name == "prediction") {
                            val timePrediction = TimePrediction()
                            timePrediction.time = xpp.getAttributeValue(2)
                            timeList.add(timePrediction)
                        }
                    }
                    eventType = xpp.next()
                }
            }.await()
            notifyObs(timeList)
        }
    }

    @Throws(XmlPullParserException::class, IOException::class)
    fun readLocation(xml: String, callback: NetworkCallback) {
        launchUI {
            val vehiculeList = VehiculeList()
            val list = async {
                val factory = XmlPullParserFactory.newInstance()
                factory.isNamespaceAware = true
                val xpp = factory.newPullParser()

                xpp.setInput(StringReader(xml))
                var eventType = xpp.eventType
                while (eventType != XmlPullParser.END_DOCUMENT) {
                    if (eventType == XmlPullParser.START_DOCUMENT) {
                    } else if (eventType == XmlPullParser.START_TAG) {
                        if (xpp.name == "vehicle") {
                            val vehicule = Vehicule()
                            vehicule.longitude = xpp.getAttributeValue(4)
                            vehicule.latitude = xpp.getAttributeValue(3)
                            vehiculeList.add(vehicule)
                        }
                    }
                    eventType = xpp.next()
                }
                vehiculeList
            }.await()
            callback.onVehicleListLoaded(list)
        }
    }

    @Throws(XmlPullParserException::class, IOException::class)
    fun readMessages(xml: String) {
        launchUI {
            var message = ""

            val factory = XmlPullParserFactory.newInstance()
            factory.isNamespaceAware = true
            val xpp = factory.newPullParser()

            xpp.setInput(StringReader(xml))
            var eventType = xpp.eventType
            while (eventType != XmlPullParser.END_DOCUMENT) {
                if (eventType == XmlPullParser.START_DOCUMENT) {
                } else if (eventType == XmlPullParser.TEXT) {
                    val text = xpp.text
                    if (text != null && !text.isEmpty() && !text.contains("\n")) {
                        message += (HtmlCompat.fromHtml(text, FROM_HTML_MODE_LEGACY).toString() + "\n \n")
                    }
                }
                eventType = xpp.next()
            }
            notifyObs(message)
        }
    }

    fun notifyObs(o: Any) {
        setChanged()
        notifyObservers(o)
    }
}
